File: src/main/java/net/frozenblock/configurableeverything/biome/util/BiomeChange.java
package net.frozenblock.configurableeverything.biome.util;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import java.util.List;

public record BiomeChange(List<BiomePlacedFeatureList> addedFeatures, List<BiomePlacedFeatureList> removedFeatures, List<BiomePlacedFeatureReplacementList> replacedFeatures, List<BiomeMusic> musicReplacements) {

	public static final Codec<BiomeChange> CODEC = RecordCodecBuilder.create(instance ->
		instance.group(
			BiomePlacedFeatureList.CODEC.listOf().fieldOf("addedFeatures").forGetter(BiomeChange::addedFeatures),
			BiomePlacedFeatureList.CODEC.listOf().fieldOf("removedFeatures").forGetter(BiomeChange::removedFeatures),
			BiomePlacedFeatureReplacementList.CODEC.listOf().fieldOf("replacedFeatures").forGetter(BiomeChange::replacedFeatures),
			BiomeMusic.CODEC.listOf().fieldOf("musicReplacements").forGetter(BiomeChange::musicReplacements)
		).apply(instance, BiomeChange::new)
	);
}

File: src/main/java/net/frozenblock/configurableeverything/biome/util/BiomeChange.kt
package net.frozenblock.configurableeverything.biome.util

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder

data class BiomeChange(
	@JvmField var addedFeatures: List<BiomePlacedFeatureList?>?,
	@JvmField var removedFeatures: List<BiomePlacedFeatureList?>?,
	@JvmField var replacedFeatures: List<BiomePlacedFeatureReplacementList?>?,
	@JvmField var musicReplacements: List<BiomeMusic?>?
) {
	companion object {
		@JvmField
		val CODEC: Codec<BiomeChange> = RecordCodecBuilder.create { instance ->
			instance.group(
				BiomePlacedFeatureList.CODEC.listOf().fieldOf("addedFeatures").forGetter(BiomeChange::addedFeatures),
				BiomePlacedFeatureList.CODEC.listOf().fieldOf("removedFeatures").forGetter(BiomeChange::removedFeatures),
				BiomePlacedFeatureReplacementList.CODEC.listOf().fieldOf("replacedFeatures").forGetter(BiomeChange::replacedFeatures),
				BiomeMusic.CODEC.listOf().fieldOf("musicReplacements").forGetter(BiomeChange::musicReplacements)
			).apply(instance, ::BiomeChange)
		}
	}

    override fun toString(): String = "BiomeChange[addedFeatures=$addedFeatures, removedFeatures=$removedFeatures, replacedFeatures=$replacedFeatures, musicReplacements=$musicReplacements]"
}
